book/src/pil/generic_declarations.md

```diff
@@ -0,0 +1,43 @@
+# Generic Declarations
+
+powdr-pil allows the same syntax to declare various kinds of symbols. This includes
+constants, fixed columns, witness columns and even macros. It deduces the symbol kind
+by its type and the way the symbol is used.
+
+You use ``let <name> = <value>;`` to declare and define a symbol. If you want to specify
+the type, you can use ``let <name>: <type> = <value>;``, but this is not yet implemented.
+
+You can omit the value which usually makes the column a witness column, but it can also
+be determined to be an intermediate column or even fully redundant.
+
+Examples:
+
+```
+let rows = 2**16; // This defines a constant
+let step = |i| i; // This defines a fixed column that contains the row number in each row.
+let x; // Here, we have a witness column.
+```
+
+## Expression Syntax
+
+Inside values for declarations, you can use a very flexible language which includes
+many different operators, function calls, lambda functions, tuple types, statement blocks,
+if and match statements and others.
+
+### Algebraic Expressions
+
+For identities (or functions called from identities), this syntax is a bit more limited:
+After inlining function calls and constant propagation, the resulting expression has to
+be an "algebraic expression". These are restricted in the following way:
+
+- You can freely use the operators  ``+``, ``-``,``*``.
+- The operator ``**`` must have a number as exponent.
+- The operator `[i]` must have a column name on the left-hand side and the index must be a number.
+- The operator `'` must have a column or `[i]` on the left-hand-side.
+- No other operators are allowed.
+
+Of course, arbitrary parentheses are allowed.
+
+### Generic Expressions
+
+TODO
```

Comment:
Is this WIP for this PR?

Reply:
no, just wanted to get something going for now.